### PR TITLE
[DC-1001] Remove Fitbit Data if Max Age Exceeded

### PR DESCRIPTION
**Pull request by dcampbell-vumc**
Created deid "cleaning rule" to remove fitbit data for all participants who are over the age of 89

### CODE REVIEW

File: data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                        sandbox_table=self.get_sandbox_tablenames()[i]),
+                cdr_consts.DESTINATION_TABLE:
+                    table,
+                cdr_consts.DESTINATION_DATASET:
+                    self.dataset_id,
+                cdr_consts.DISPOSITION:
+                    WRITE_TRUNCATE
+            })
+
+        return [sandbox_queries_list, drop_queries_list]
```
[the remark this answers is not in the document]
Comment (dcampbell-vumc) [reply data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:127]: Done, used `return sandbox_queries_list.extend(drop_queries_list)` had to also change the assert statements in the unit test.

File: data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+        Set the issue numbers, description and affected datasets. As other tickets may affect
+        this SQL, append them to the list of Jira Issues.
+        DO NOT REMOVE ORIGINAL JIRA ISSUE NUMBERS!
+        """
+        desc = (
+            'Drops all FitBit data from participants whose max age exceeds 89')
+        super().__init__(issue_numbers=['DC1001'],
+                         description=desc,
+                         affected_datasets=[cdr_consts.RDR],
```
[the remark this answers is not in the document]
Comment (dcampbell-vumc) [reply data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:82]: > I think we should add a new stage to this Fitbit dataset cleaning.

@ksdkalluri I don't understand/know how to do that. Can you explain?

[the remark this answers is not in the document]
Comment (dcampbell-vumc) [reply data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:82]: Do I need to add something to `clean_cdr.py`?

File: data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+SAVE_ROWS_TO_BE_DROPPED_QUERY = jinja_env.from_string("""
+CREATE OR REPLACE TABLE `{{project}}.{{sandbox_dataset}}.{{sandbox_table}}` AS
+SELECT * FROM `{{project}}.{{dataset}}.{{table}}`
+WHERE person_id IN (
+  SELECT person_id
+    FROM (
+    SELECT DISTINCT person_id,
+        EXTRACT(YEAR FROM CURRENT_DATE()) - year_of_birth AS age
+            FROM `{{project}}.{{dataset}}.person` ORDER BY 2)
+        WHERE age > 89)
```
Comment (lrwb-aou) [inline data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:51]: @dcampbell-vumc This is not going to be the exact same list as what is used by deid.  So, there will potentially be a small subset of people in this table that do not make it into deid.  


You are sandboxing anyone older than 89.  `curation/data_steward/tools/run_deid.py` is using `age < {{max_age}}` where max_age is a parameter.  So, your sandbox won't include those people who are 89 and they will be included into the released fitbit data.  deid will not include these people in deid tables.  So they don't need to be present here.

Comment (dcampbell-vumc) [reply data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:51]: Should I parameterize max age? I went ahead and changed it to `WHERE age >= 89` and tested it to make sure that those who are exactly 89 are sandboxed and dropped, so it won't be too much more work to parameterize max age.

Comment (ksdkalluri) [reply data_steward/cdr_cleaner/cleaning_rules/deid/remove_fitbit_data_if_max_age_exceeded.py:51]: Thanks @lrwb-aou for finding this. I didn't cross-check this with current deid.

File: data_steward/cdr_cleaner/clean_cdr.py [hunk 4]
```diff
@@ -259,6 +264,18 @@ def _gather_unioned_ehr_queries(project_id, dataset_id, sandbox_dataset_id):
                            sandbox_dataset_id)
 
 
+def _gather_fitbit_cleaning_queries(project_id, dataset_id, sandbox_dataset_id):
```
Comment (ksdkalluri) [inline data_steward/cdr_cleaner/clean_cdr.py:267]: This function is not callable without the following missing code  

```
def get_fitbit_dataset_id():
  return os.environ.get('FITBIT_DATASET_ID')


def clean_fitbit_dataset(project_id=None, dataset_id=None):
    """
    Run all clean rules defined for the Fitbit dataset.

    :param project_id:  Name of the BigQuery project.
    :param dataset_id:  Name of the dataset to clean
    """
    if project_id is None:
        project_id = app_identity.get_application_id()
        LOGGER.info(
            f"Project is unspecified.  Using default value of:\t{project_id}")

    if dataset_id is None:
        dataset_id = get_fitbit_dataset_id()
        LOGGER.info(
            f"Dataset is unspecified.  Using default value of:\t{dataset_id}")

    sandbox_dataset_id = sandbox.create_sandbox_dataset(project_id=project_id,
                                                        dataset_id=dataset_id)

    query_list = _gather_fitbit_cleaning_queries(project_id, dataset_id,
                                                    sandbox_dataset_id)

    LOGGER.info("Cleaning FITBIT dataset")
    clean_engine.clean_dataset(project_id, query_list, stage.FITBIT)
```

Comment (dcampbell-vumc) [reply data_steward/cdr_cleaner/clean_cdr.py:267]: I added the functions 👍🏾

**Pull request merged**